engines/cryomni3d/versailles/documentation.cpp

```diff
@@ -0,0 +1,2070 @@
+
+	while (true) {
+		currentPos = getDocPartAddress(currentPos, end, patterns);
+		if (!currentPos) {
+			break;
+		}
+		currentPos -= 5;
+		if (first) {
+			record.position = currentPos - allDocs;
+			record.id = 0;
+
+			lastRecordName = currentPos + 5;
+
+			first = false;
+		} else {
+			record.size = (currentPos - allDocs) - record.position;
+			//debug("Found record %s", lastRecordName);
+			_records[lastRecordName] = record;
+			_recordsOrdered.push_back(lastRecordName);
+
+			record.id++;
+			record.position = currentPos - allDocs;
+
+			lastRecordName = currentPos + 5;
+		}
+		// Next line
+		currentPos += strlen(currentPos) + 1;
+	}
+	record.size = allDocsSize - record.position;
+	_records[lastRecordName] = record;
+	_recordsOrdered.push_back(lastRecordName);
+
+	delete[] allDocs;
+}
+
+void Versailles_Documentation::handleDocArea() {
+	g_system->showMouse(false);
+
+	// Load all links lazily and free them at the end to not waste memory
+	// Maybe it's not really useful
+	getLinks("ALL00", _allLinks);
+
+	bool end = false;
+	while (!end) {
+		Common::String selectedRecord = docAreaHandleSummary();
+		if (selectedRecord == "") {
+			end = true;
+		} else if (selectedRecord == "VT00") {
+			selectedRecord = docAreaHandleTimeline();
+			if (selectedRecord != "") {
+				if (docAreaHandleRecords(selectedRecord) == 2) {
+					end = true;
+				}
+			}
+		} else {
+			if (docAreaHandleRecords(selectedRecord) == 2) {
+				end = true;
+			}
+		}
+	}
+
+	_allLinks.clear();
+
+	g_system->showMouse(true);
+}
+
+void Versailles_Documentation::handleDocInGame(const Common::String &record) {
+	_visitTrace.clear();
+	_currentRecord = record;
+
+	Graphics::ManagedSurface docSurface;
+	Common::String nextRecord;
+	MouseBoxes boxes(3);
+
+	g_system->showMouse(false);
+	bool end = false;
+	while (!end) {
+		inGamePrepareRecord(docSurface, boxes);
+		unsigned int action = inGameHandleRecord(docSurface, boxes, nextRecord);
+		switch (action) {
+		case 0:
+			// Back
+			if (!_visitTrace.empty()) {
+				_currentRecord = _visitTrace.back();
+				_visitTrace.pop_back();
+				break;
+			}
+		// No previous record, like a quit
+		// fall through
+		case 1:
+			// Quit
+			end = true;
+			break;
+		case 2:
+			// Follow hyperlink keeping trace
+			_visitTrace.push_back(_currentRecord);
+			_currentRecord = nextRecord;
+			break;
+		default:
+			error("Invalid case %d when displaying doc record", action);
+		}
+	}
+	g_system->showMouse(true);
+}
+
+Common::String Versailles_Documentation::docAreaHandleSummary() {
+	struct Category {
+		const char *record;
+		const char *bmp;
+		Common::Point imgPos;
+		Common::Rect linesPos;
+		const Common::String *title;
+		Graphics::Surface highlightedImg;
+
+		Category(const char *record_, const char *bmp_, const Common::Point &imgPos_,
+		         const Common::Rect &linesPos_, const Common::String *title_) :
+			record(record_), bmp(bmp_), imgPos(imgPos_), linesPos(linesPos_), title(title_) { }
+	} categories[8] = {
```

Comment:
make it const?

Reply:
I can't make it const because there is a Surface in each object which gets written below when loading the image file